check-meta: wrap maintainers attribute to include team members

SigmaSquadron opened the pull request:
This is a work-around that resolves the issues that spawned from #400458 by adding the team members from teams declared in `meta.teams` to `meta.maintainers`, effectively restoring the original behaviour of the `maintainers` attribute, and thus allowing downstreams to keep consuming `meta.maintainers` without any extra changes to support `meta.teams`.

Follow-up to #394797.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

flokli commented:
Please add reasoning both in the commit message(s) and PR description.

JohnRTitor commented:
At this point, I feel like, meta.teams should and should only be a wrapper for meta.maintainers, so it doesn't break existing functionality.

adding a team to meta.teams, would add their members to meta.maintainers. This should maintain backwards compatibility and would be compatible with repology as it is.

CC @RossComputerGuy

numinit commented:
I believe why it was done the existing way is that meta is mostly a dumb passthrough on mkDerivation. Fully agreed, though, if it can be done this way it would be preferable and would make sure everything is OK downstream.

mweinelt commented:
> Maintainers wasn't explicitly a formally stable API, but we should treat it that way at this point).

If you don't specify what is part of your API and what isn't then everything becomes your API. That actually happens a lot with nixpkgs.

SigmaSquadron committed:
check-meta: wrap maintainers attribute to include team members

This is a work-around that resolves the issues that spawned from #400458 by adding the team members from teams declared in `meta.teams` to `meta.maintainers`, effectively restoring the original behaviour of the `maintainers` attribute, and thus allowing downstreams to keep consuming `meta.maintainers` without any extra changes to support `meta.teams`.

Follow-up to #394797.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
> Instead of merging this one, @SigmaSquadron perhaps we can open a PR doing what has been suggested and discussed above?

I'm all for it! Unfortunately, I'm not entirely sure how to do that. My first instinct was adding [these](https://github.com/SigmaSquadron/nixpkgs/blob/baa360f71982a31bdc0bc92d382ca76b36f4efa5/pkgs/stdenv/generic/check-meta.nix#L597-L599) lines to `check-meta.nix`.

numinit commented:
We should remove this consumer that was added in #394797: https://github.com/NixOS/nixpkgs/blob/master/ci/eval/compare/maintainers.nix#L58

numinit reviewed (commented):
Is the hasNoMaintainers predicate in check-meta now redundant, or do we still need to check both maintainers and teams?

RossComputerGuy commented:
I think leave it, I don't think it passes through in this state.

SigmaSquadron commented:
> We should remove this consumer that was added in #394797: https://github.com/NixOS/nixpkgs/blob/master/ci/eval/compare/maintainers.nix#L58
> 
> Apart from that, we can close the Repology and s.n.o. PRs, since the behavior is now compatible with both of them.

I was hoping this workaround would be temporary, until all consumers of meta.maintainers adopt the new teams attribute.

RossComputerGuy commented:
That's what I was thinking as well.

numinit commented:
OK. Regardless we should just make sure that we don't duplicate pings, because it will (in the current state).

Long term I definitely agree with s.n.o listing teams. However, I don't think Repology much cares. The maintainers attribute is an external API contract that we have effectively created with them to display maintainers, so "maintainers are just inclusive of members of all teams" is the way to avoid breaking it. And we should generally try to avoid breaking external API contracts.

Speaking of which, we may want to also deduplicate maintainers in the eval in case someone is on more than one team maintaining something.

SigmaSquadron commented:
We could wrap everything with `lib.unique`. It may slow down eval though; I'm not sure how efficient that function is.

numinit commented:
Fair point. It is likely not necessary. (I think what we can easily dedup is more important anyway, like the eval script, since currently I think it's adding teams to maintainers which already take the teams into account?)

numinit commented:
> We should remove this consumer that was added in #394797: https://github.com/NixOS/nixpkgs/blob/master/ci/eval/compare/maintainers.nix#L58

I think this still applies. At present it'd be concatenating maintainers (which are inclusive of teams) with all the teams.

mweinelt commented:
It's been a week. I'm not getting repology notifications for a number of packages any more. This is breaking my workflow.

Can we get on with this?

numinit committed:
ci/eval: use correct maintainer and team positions

We need to pass through the maintainers and teams positions from the
original meta so pings work correctly, since check-meta clobbers the
original attribute positions in them.

Tested with `maintainers/scripts/get-maintainer-pings-between.sh` on a
handful of major packages maintained by both individuals and teams.

numinit commented:
Did some prodding at it tonight to make sure this does not break anything that depends on it (notably, maintainer pings). Also added a comment asking to please not break this again since it is an external API and Repology depends on it.

numinit commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -589,11 +593,24 @@ let
+      # Maintainers should be inclusive of teams.
+      # Note that there may be external consumers of this API (repology, for instance) -
+      # if you add a new maintainer or team attribute please ensure that this expectation is still met.
+      maintainers =
+        attrs.meta.maintainers or [ ]
+        ++ concatMap (team: team.members or [ ]) attrs.meta.teams or [ ];
```
I saw it come up in a couple of the places I tried. Another idea is a listToAttrs, but that'll probably be slightly heavier on memory since we'll just throw out the attrset anyway. May be worth a shot, though, since it's not N^2 at least?

numinit replied:
Now that Nix has tracing support, it'll be interesting to use it to test dedup while computing meta.

winterqt commented on ci/eval/compare/maintainers.nix:
```diff
@@ -64,7 +62,8 @@ let
       builtins.map (pos: lib.strings.removePrefix (toString ../..) pos.file) (
         builtins.filter (x: x != null) [
-          (builtins.unsafeGetAttrPos "maintainers" (drv.meta or { }))
+          ((drv.meta or { }).maintainersPosition or null)
+          ((drv.meta or { }).teamsPosition or null)
```
Why not

```suggestion
          (builtins.unsafeGetAttrPos "maintainers" (drv.meta or { }))
          (builtins.unsafeGetAttrPos "teams" (drv.meta or { }))
```

? Ideally we wouldn't add (slight!) extra eval overhead just for CI scripts (which AFAIU check-meta will do).

numinit replied:
The new maintainers attribute was actually clobbering the original one's position, so it would return check-meta.nix every time rather than the real declaring file. I needed a way to pull the location of the original maintainers (pre-teams concatenation) out of there, and just did the same with teams. If it's written this way, it'll return a position pointing to check-meta.nix.

check-meta.nix is a little counterintuitive, but it's a bunch of // attrset updates that both prefers certain things in the meta and overrides other things in it. Maintainers is written as one of the overrides so I made sure we're holding onto the original.

winterqt reviewed (approved):
Thanks! :)

winterqt commented:
(Why remove from the stdenv board?)

Also, I didn't know we backported this change to 24.11. oof.

SigmaSquadron commented:
I removed it from Stdenv and CUDA because this isn't part of their projects. The projects are added automatically if we touch certain files during treewides.

We also didn't backport this; it was another mishap caused by changing files in the `ci/` directory.

winterqt commented:
Makes sense.

Ah, right, forgot about that. Sorry!

numinit commented:
Thanks for the work and the merge, y'all.
